### PR TITLE
Handle assertThat in JUnit and Hamcrest.

### PR DESCRIPTION
**Pull request by navahgar**
* This adds handling of nullness assertions using "assertThat(..., isNotNull())" statements from JUnit (<https://junit.org/junit4/javadoc/latest/org/junit/Assert.html>) and Hamcrest (https://junit.org/junit4/javadoc/4.12/org/hamcrest/MatcherAssert.html).
* Issue #301

### CODE REVIEW

File: nullaway/src/main/java/com/uber/nullaway/handlers/AssertionHandler.java [hunk 1]
```diff
@@ -52,6 +62,17 @@
   private Name assertThat;
   private Name assertThatOwner;
 
+  // Names for junit assertion libraries.
+  private Name hamcrestAssertClass;
+  private Name junitAssertClass;
+
+  // Names for matchers.
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/handlers/AssertionHandler.java]: "for hamcrest matchers"?

File: nullaway/src/main/java/com/uber/nullaway/handlers/AssertionHandler.java [hunk 2]
```diff
@@ -97,6 +132,50 @@ private boolean isMethodAssertThat(Symbol.MethodSymbol methodSymbol) {
...
+    if (matchesMatcherMethod(node, notMatcher)) {
+      // All overloads of `not` method have exactly one argument.
+      return isMatcherNull(((MethodInvocationNode) node).getArgument(0));
+    }
+    return matchesMatcherMethod(node, notNullValueMatcher);
+  }
+
+  private boolean isMatcherNull(Node node) {
+    // Matches with nullValue()
+    return matchesMatcherMethod(node, nullValueMatcher);
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/handlers/AssertionHandler.java]: Does this also need a case for `not(...)`. Is `assertThat(o, is(not(not(notNullValue()))))` valid code? Do we care?

File: nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java [hunk 3]
```diff
@@ -1030,10 +1030,74 @@ public void doNotSupportAssertThatWhenDisabled() {
             "import java.util.Objects;",
             "import javax.annotation.Nullable;",
             "class Test {",
-            "  private void foo(@Nullable Object o) {",
-            "    com.google.common.truth.Truth.assertThat(o).isNotNull();",
+            "  private void foo(@Nullable Object a, @Nullable Object b, @Nullable Object c) {",
+            "    com.google.common.truth.Truth.assertThat(a).isNotNull();",
```
Comment (lazaroclapp) [inline nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: Can we separate the Truth test cases from the JUnit/Hamcrest ones? I think also that way we can properly use imports and make them far clearer/more idiomatic to read.

Comment (lazaroclapp) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: Oh. I see we are doing that for the "enabled" case. I still think it's worth to split the "assert handler disabled" tests too, but not super set on it.

File: nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java [hunk 4]
```diff
@@ -1030,10 +1030,74 @@ public void doNotSupportAssertThatWhenDisabled() {
...
+                "-XepOpt:NullAway:HandleTestAssertionLibraries=true"))
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import java.lang.Object;",
+            "import java.util.Objects;",
+            "import javax.annotation.Nullable;",
+            "class Test {",
+            "  private void foo(@Nullable Object a, @Nullable Object b) {",
+            "    org.hamcrest.MatcherAssert.assertThat(a, org.hamcrest.Matchers.is("
```
Comment (lazaroclapp) [inline nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: Just add the static imports to make this both easier to read and more representative of idiomatic code:

```
import static org.hamcrest.MatcherAssert.assertThat;
...
assertThat(a, is(notNullValue));
```

File: nullaway/src/main/java/com/uber/nullaway/handlers/AssertionHandler.java (hunk 2 above)
Comment (navahgar) [reply nullaway/src/main/java/com/uber/nullaway/handlers/AssertionHandler.java]: It should be pretty bad code in that case. If ever that comes up, I think we could safely ask the users to remove the double negation here for it to work.

If for some weird reason this becomes a necessity, it wouldn't be too hard to add this. I just don't want to add it without knowing if there is a real reason to write code like this.

Comment (lazaroclapp) [reply nullaway/src/main/java/com/uber/nullaway/handlers/AssertionHandler.java]: Fair enough. I agree :)

File: nullaway/src/main/java/com/uber/nullaway/handlers/AssertionHandler.java (hunk 1 above)
Comment (navahgar) [reply nullaway/src/main/java/com/uber/nullaway/handlers/AssertionHandler.java]: Done.

File: nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java (hunk 3 above)
Comment (navahgar) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: Separated the negative tests as well.

File: nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java (hunk 4 above)
Comment (navahgar) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: Sure, good point. I didn't know about static imports. Updated the tests now.

**Comment by lazaroclapp** [comment]
Looks great to me. Feel free to "Squash and merge" after CI passes :)

**Pull request merged**